### PR TITLE
[fix](nereids)fix some nereids bugs

### PR DESCRIPTION
**Pull request by starocean999**
# Proposed changes

Issue Number: close #xxx

## Problem summary

1.remove forcing nullable for slot on EmptySetNode.
2.order by xxx desc should use nulls last as default order.
3.don't create runtime filter if runtime filter mode is OFF.
4.group by constant value need check the corresponding expr shouldn't have any aggregation functions.
5.fix two left outer join reorder bug( A left join B left join C).
6.fix semi join and left outer join reorder bug.( A left join B semi join C ).
7.fix group by NULL bug.
8.change ceil and floor function to correct signature.
9.add literal comparasion for string and date type.
10.fix the getOnClauseUsedSlots method may not return valid value.
11.the tightness common type of string and date should be date.
12.the nullability of set operation node's result exprs is not set correctly.
13.Sort node should remove redundent ordering exprs.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.93 seconds
 load time: 479 seconds
 storage size: 17121594761 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230108142350_clickbench_pr_75649.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 1]
```diff
@@ -1102,10 +1102,13 @@ public PlanFragment visitPhysicalProject(PhysicalProject<? extends Plan> project
         }
 
         TupleDescriptor tupleDescriptor = generateTupleDesc(slotList, null, context);
-        inputPlanNode.setProjectList(execExprList);
+        if (!(inputPlanNode instanceof EmptySetNode)) {
+            inputPlanNode.setProjectList(execExprList);
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:1107]: add a todo, we need to merge project(emptyset) to a new emptyset in rewrite step of Nereids

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java [hunk 2]
```diff
@@ -74,6 +76,11 @@ public List<Rule> buildRules() {
                                     int ord = i.getIntValue();
                                     checkOrd(ord, aggOutput.size());
                                     Expression aggExpr = aggOutput.get(ord - 1);
+                                    if (aggExpr.containsType(AggregateFunction.class)) {
+                                        throw new AnalysisException(
+                                                "GROUP BY expression must not contain aggregate functions: "
+                                                        + aggExpr.toSql());
+                                    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java:83]: we should not add check here, instead we need to add a check in CheckAnalysis

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/NormalizeToSlot.java [hunk 3]
```diff
@@ -133,7 +136,10 @@ public static NormalizeToSlotTriplet toTriplet(Expression expression, @Nullable
             }
 
             Alias alias = new Alias(expression, expression.toSql());
-            return new NormalizeToSlotTriplet(expression, alias.toSlot(), alias);
+            SlotReference slot = (SlotReference) alias.toSlot();
+            // BE will create a nullable uint8 column to expand NullLiteral when necessary
+            return new NormalizeToSlotTriplet(expression, expression instanceof NullLiteral ? slot.withDataType(
+                    TinyIntType.INSTANCE) : slot, alias);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/NormalizeToSlot.java:142]: if we just want to process `group by null`, maybe a better way to do it is generating NullLiteral with TinyIntType when we do rewrite.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 4]
```diff
@@ -151,11 +152,14 @@ public static Pair<List<ExprId>, List<ExprId>> getOnClauseUsedSlots(
 
         for (Expression expr : join.getHashJoinConjuncts()) {
             EqualTo equalTo = (EqualTo) expr;
-            if (!(equalTo.left() instanceof Slot) || !(equalTo.right() instanceof Slot)) {
+            Optional<Slot> leftSlot = ExpressionUtils.extractSlotOrCastOnSlot(equalTo.left());
+            Optional<Slot> rightSlot = ExpressionUtils.extractSlotOrCastOnSlot(equalTo.right());
+            if (!leftSlot.isPresent() || !rightSlot.isPresent()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:157]: add a TODO on it, to explain why and we will fix normalize join hash equals future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java [hunk 5]
```diff
@@ -215,6 +215,9 @@ public static Optional<DataType> findTightestCommonType(BinaryOperator binaryOpe
             }
         } else if (left instanceof CharacterType && right instanceof CharacterType) {
             tightestCommonType = CharacterType.widerCharacterType((CharacterType) left, (CharacterType) right);
+        } else if (left instanceof CharacterType && right instanceof DateLikeType) {
+            // TODO: need check implicitCastMap to keep the behavior consistent with old optimizer
+            tightestCommonType = right;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:220]: should we also add `right instanceof CharacterType && left instanceof DateLikeType`

File: fe/fe-core/src/main/java/org/apache/doris/planner/SortNode.java [hunk 6]
```diff
@@ -320,7 +320,12 @@ public Set<SlotId> computeInputSlotIds(Analyzer analyzer) throws NotImplementedE
      */
     public void finalizeForNereids(TupleDescriptor tupleDescriptor,
             List<Expr> outputList, List<Expr> orderingExpr) {
-        resolvedTupleExprs = Lists.newArrayList(orderingExpr);
+        resolvedTupleExprs = Lists.newArrayList();
+        for (Expr order : orderingExpr) {
+            if (!resolvedTupleExprs.contains(order)) {
+                resolvedTupleExprs.add(order);
+            }
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/SortNode.java:328]: add a TODO, we need to fix it in Nereids' code later

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java [hunk 7]
```diff
@@ -74,6 +76,12 @@ public List<Rule> buildRules() {
                                     int ord = i.getIntValue();
                                     checkOrd(ord, aggOutput.size());
                                     Expression aggExpr = aggOutput.get(ord - 1);
+                                    if (aggExpr instanceof Alias) {
+                                        aggExpr = ((Alias) aggExpr).child();
+                                        if (aggExpr instanceof Literal) {
+                                            aggExpr = groupByExpr;
+                                        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java:83]: we should replaced by expression in output list in any case

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/SemiJoinLogicalJoinTranspose.java [hunk 8]
```diff
@@ -83,12 +85,17 @@ public Rule build() {
...
-                                a, c);
-                        return new LogicalJoin<>(bottomJoin.getJoinType(), bottomJoin.getHashJoinConjuncts(),
-                                bottomJoin.getOtherJoinConjuncts(), JoinHint.NONE, newBottomSemiJoin, b);
+                        if (bottomJoinType.isInnerJoin() || bottomJoinType.isRightOuterJoin()) {
+                            LogicalJoin<GroupPlan, GroupPlan> newBottomSemiJoin = new LogicalJoin<>(
+                                    topSemiJoin.getJoinType(),
+                                    topSemiJoin.getHashJoinConjuncts(), topSemiJoin.getOtherJoinConjuncts(),
+                                    JoinHint.NONE,
+                                    a, c);
+                            return new LogicalJoin<>(JoinType.INNER_JOIN, bottomJoin.getHashJoinConjuncts(),
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/SemiJoinLogicalJoinTranspose.java:94]: return new LogicalJoin<>(bottomJoinType, bottomJoin.getHashJoinConjuncts(),

Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/SemiJoinLogicalJoinTranspose.java:94]: return new LogicalJoin<>(bottomJoinType,

Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/SemiJoinLogicalJoinTranspose.java:94]: return new LogicalJoin<>(bottomJoinType,

**Review by morrySnow — APPROVED** [review]

**Pull request merged**